File: src/142.linked-list-cycle-ii/solution_two-pointers.ts
/*
 * @lc app=leetcode.cn id=142 lang=typescript
 *
 * https://leetcode-cn.com/problems/linked-list-cycle-ii/
 */

/**
 * Definition for singly-linked list.
 */
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

// @lc code=start
function findFirstEncounter(head: ListNode | null) {
  if (!head) {
    return null;
  }

  let slower: ListNode | null = head;
  let faster: ListNode | null = head;

  while (faster) {
    slower = (slower as ListNode).next;
    faster = faster.next ? faster.next.next : null;

    if (slower && slower === faster) {
      return slower;
    }
  }

  return null;
}

function findCycleEntry(head: ListNode | null) {
  const encounter = findFirstEncounter(head);

  if (!encounter) {
    return null;
  }

  let first = head as ListNode;
  let second: ListNode | null = encounter;

  while (second) {
    if (first === second) {
      return first;
    }

    first = first.next as ListNode;
    second = second.next;
  }

  return null;
}

function detectCycle(head: ListNode | null): ListNode | null {
  return findCycleEntry(head);
}
// @lc code=end

export default detectCycle;

File: src/142.linked-list-cycle-ii/solution_two-pointers.cpp
/*
 * @lc app=leetcode.cn id=142 lang=cpp
 *
 * https://leetcode-cn.com/problems/linked-list-cycle-ii/
 */

/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
};

// @lc code=start

class Solution {
 private:
  ListNode *findFirstEncounter(ListNode *head) {
    if (!head) {
      return nullptr;
    }

    ListNode *slower = head;
    ListNode *faster = head;

    while (faster) {
      slower = slower->next;
      faster = faster->next ? faster->next->next : nullptr;

      if (faster && faster == slower) {
        return faster;
      }
    }

    return nullptr;
  }

  ListNode *findCycleEntry(ListNode *head) {
    ListNode *encounter = findFirstEncounter(head);

    if (!encounter) {
      return nullptr;
    }

    ListNode *slower = head;
    ListNode *faster = encounter;

    while (faster) {
      if (faster == slower) {
        return faster;
      }

      slower = slower->next;
      faster = faster->next;
    }

    return nullptr;
  }

 public:
  ListNode *detectCycle(ListNode *head) { return findCycleEntry(head); }
};

// @lc code=end
